cylc/flow/dbstatecheck.py

```diff
@@ -274,34 +281,50 @@ def workflow_state_query(
                     res.append(fstr)
             db_res.append(res)
 
-        if not (is_output or is_message):
+        if target_table == CylcWorkflowDAO.TABLE_TASK_STATES:
             return db_res
 
+        warn_output_fallback = is_output
         results = []
         for row in db_res:
-            outputs_map = json.loads(row[2])
-            if is_message:
-                # task message
-                try:
-                    outputs = list(outputs_map.values())
-                except AttributeError:
-                    # Cylc 8 pre 8.3.0 back-compat: list of output messages
-                    outputs = list(outputs_map)
+            outputs: Union[Dict[str, str], List[str]] = json.loads(row[2])
+            if isinstance(outputs, dict):
+                messages: Iterable[str] = outputs.values()
             else:
-                # task output
-                outputs = list(outputs_map)
+                # Cylc 8 pre 8.3.0 back-compat: list of output messages
+                messages = outputs
+                if warn_output_fallback:
+                    LOG.warning(output_fallback_msg)
+                    warn_output_fallback = False
```

Comment:
Clearer distinction between messages and outputs, added warning for 8.0.0-8.3.0 tasks